src/Numeric/Units/Dimensional/Dynamic.hs

```diff
@@ -101,19 +101,23 @@ instance Num a => Monoid (AnyQuantity a) where
 -- not be a 'Quantity' of any 'Dimension' whatsoever, but in exchange it gains instances
 -- for the common numeric classes. It's therefore useful for manipulating, and not merely storing,
 -- quantities of unknown dimension.
-newtype DynQuantity a = DynQuantity (Maybe (AnyQuantity a))
-  deriving (Eq, Data, Generic, Generic1, Typeable, Show)
+data DynQuantity a = GoodQuantity (AnyQuantity a) | BadQuantity
+  deriving (Data, Generic, Generic1, Typeable, Show)
+
+instance Eq a => Eq (DynQuantity a) where
```

Comment:
Ah, to be honest I am not sure. I was thinking more in analogy to `NaN != NaN`, but I guess that is an IEEE 754 thing which perhaps one shouldn't read to much into.

The case I fear is when you, e.g.,  perform some arithmetic on two different inputs and branch on their equality, and unintentionally and unexpectedly go down the `True` branch due to them having bad dimensions rather than them being equal.

(Of course, the right thing to do is to `promoteQuantity` before performing the equality test. By the way, is there a use case for promoting from `DynQuantity` only to `AnyQuantity`?)